### PR TITLE
sway: Set bar defaults to null

### LINKED ISSUE #1361: wayland.windowManager.sway.config.bars has weird defaults
### Issue description
My bar configuration for sway is very simple:
```
bar {
    position top
    swaybar_command waybar
}
```
But setting this with home-manager as such:
```
bars = [
  {
    position = "top";
    command = "waybar";
  }
];
```
Results in:
```
bar {

  font pango:monospace 8
  mode dock
  hidden_state hide
  position top
  status_command /nix/store/h7s6i9q1z5fxrlyyw5ls8vqxhf5bcs5a-i3status-2.13/bin/i3status
  swaybar_command waybar
  workspace_buttons yes
  strip_workspace_numbers no
  tray_output primary
  colors {
    background #000000
    statusline #ffffff
    separator #666666
    focused_workspace #4c7899 #285577 #ffffff
    active_workspace #333333 #5f676a #ffffff
    inactive_workspace #333333 #222222 #888888
    urgent_workspace #2f343a #900000 #ffffff
    binding_mode #2f343a #900000 #ffffff
  }

}
```

The defaults don't really make sense. `i3status` should not be pulled and used for swaywm. And in the usecase of waybar, all the other settings are unnecessary.

#### Maintainer CC

@alexarice 
#### Technical details

 - system: `"x86_64-linux"`
 - host os: `Linux 5.4.48, NixOS, 20.03.2351.f8248ab6d9e (Markhor)`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.3.6`
 - channels(root): `"nixos-20.03.2351.f8248ab6d9e"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`

### PR DESCRIPTION
**Pull request by alexarice**
### Description

Fixes #1361
Allows most of the fields in `bar` to be null. The defaults bar then sets these. This should mean that the default bar is unchanged but if the user creates a new bar with
```
bar {
  command = x;
}
```
then there won't be lots of unnecessary fields anymore. These changes are guarded by a stateVersion.

### Checklist

- [x] Change is backwards compatible.

- [x] Code formatted with `./format`.

- [ ] Code tested through `nix-shell --pure tests -A run.all`. 
The tests run fine but is there a way to test this with `stateVersion = "20.09"` @rycee 

- [ ] Test cases updated/added. See [example](https://github.com/rycee/home-manager/commit/f3fbb50b68df20da47f9b0def5607857fcc0d021#diff-b61a6d542f9036550ba9c401c80f00ef).
No changes atm though would be nice to add separate tests for pre 20.09 and post 20.09

- [x] Commit messages are formatted like

    ```
    {component}: {description}

    {long description}
    ```

    See [CONTRIBUTING](https://github.com/rycee/home-manager/blob/master/doc/contributing.adoc#sec-commit-style) for more information and [recent commit messages](https://github.com/rycee/home-manager/commits/master) for examples.

- If this PR adds a new module

  - [ ] Added myself as module maintainer. See [example](https://github.com/rycee/home-manager/blob/068ff76a10e95820f886ac46957edcff4e44621d/modules/programs/lesspipe.nix#L6).

  - [ ] Added myself and the module files to `.github/CODEOWNERS`.

### CODE REVIEW

File: modules/services/window-managers/i3-sway/lib/options.nix [hunk 1]
```diff
@@ -49,8 +50,21 @@ let
   };
 
   barModule = types.submodule {
-    options = {
-      inherit fonts;
+    options = let
+      newVersion = versionAtLeast config.home.stateVersion "20.09";
+      nullableOption = { type, default, description }:
```
Comment (rycee) [inline modules/services/window-managers/i3-sway/lib/options.nix:55]: Maybe `mkNullableOption` to fit the theme of `mkOption` and `mkEnableOption`?

Comment (rycee) [inline modules/services/window-managers/i3-sway/lib/options.nix:55]: Could use `@ args` to simplify and generalize the code. Something like
```nix
{ type, default, description, ... } @ args:
  if newVersion then
    mkOption (args // {
      type = types.nullOr type;
      default = null;
      example = default;
    })
  else
    mkOption args
```

File: modules/services/window-managers/i3-sway/lib/options.nix [hunk 2]
```diff
@@ -49,8 +50,21 @@ let
   };
 
   barModule = types.submodule {
-    options = {
-      inherit fonts;
+    options = let
+      newVersion = versionAtLeast config.home.stateVersion "20.09";
```
Comment (rycee) [inline modules/services/window-managers/i3-sway/lib/options.nix:54]: `newVersion` is a bit vague, perhaps `atLeast2009` or something?

File: modules/services/window-managers/i3-sway/lib/options.nix (hunk 1 above)
Comment (rycee) [reply modules/services/window-managers/i3-sway/lib/options.nix:55]: Or even more simplerized
```nix
{ type, default, description, ... } @ args:
  mkOption (args //
    optionalArgs newVersion {
      type = types.nullOr type;
      default = null;
      example = default;
    })
```

**Comment by alexarice** [comment]
This is more a note to myself but this will need a changelog entry somewhere right?

**Comment by rycee** [comment]
Yeah, it goes in the "State Version Changes" section here: https://github.com/rycee/home-manager/blob/master/doc/release-notes/rl-2009.adoc#state-version-changes

**Comment by alexarice** [comment]
Have tried to fix up everything and add documentation though have hit 2 problems.
- Is there a way to coerce any nix object to a string in `defaultText`. I effectively want to mention the old default but at the moment have just put `see code`
- The `stateVersion` changelog references an option that does not exist on macos. Any way to fix this?

**Comment by rycee** [comment]
I guess you can use any number of solutions to produce the default string. Perhaps `generators.toPretty`, copy the text manually, make your own function to produce the string, or something similar? Perhaps something basic like
```
attrsStr =
  let
    mkPrefix = prefix: n: "${optionalString (prefix != "") "${prefix}."}${n}";
    go = prefix: mapAttrsToList (n: v: (if isAttrs v then go else nameValuePair) (mkPrefix prefix n) v);
  in attrs: generators.toKeyValue {} (listToAttrs (flatten (go "" attrs)))
```
would work?

If you can't reference the option due to macOS then I would suggest just removing the reference.

**Comment by alexarice** [comment]
@rycee I have fixed this up and added a test, At the moment I have wrapped the options that don't appear in macos in quote marks `'` but there might be a better way (I am not really fluent in asciidoc)

**Comment by rycee** [comment]
Thanks! LGTM. Rebased to master in 0869e237007aa48f458c9d9ba9080280565a7160.

**Pull request closed without merging**